### DISCUSSION TITLE
Suggestion for a new InputModeAttribute attribute in Data Annotations

### DISCUSSION THREAD

**Discussion by Andres10121997** [dotnet/runtime#108150] [Ideas]
Hello ASP.NET team,
I would like to propose the addition of a new Data Annotation attribute called `InputModeAttribute`, which would allow you to specify the HTML attribute `inputmode` in models. This functionality would improve the usability of forms in web applications, providing developers with an easy way to control keyboard behavior on mobile devices.

# Attribute Details
The `InputModeAttribute` would allow developers to define the input mode to be used in a specific field. This is especially useful in fields such as email, numbers, and searches.

# Proposed Enumeration
```
public class InputModeAttribute : ValidationAttribute
{
    public And InputModeen
    {
        @decimal,
        email,
        none,
        numeric,
        search,
        tel
        text,
        url
    }

    public InputModeEnum InputMode { get; }

    public InputModeAttribute(InputModeEnum inputMode)
    {
        InputMode = inputMode;
    }

    public string GetInputMode() => InputMode.ToString();
}
```

# Example of Use
Developers could use the new attribute in their models as follows:
```
public class MyModel
{
    [Required]
    [InputMode(InputModeAttribute.InputModeEnum.email)]
    public string Email { get; set; }

    [Required]
    [InputMode(InputModeAttribute.InputModeEnum.numeric)]
    public string Edad { get; set; }
}
```

# Proceeds
1. **Improved Usability:** It will facilitate data entry on mobile devices by displaying appropriate keyboards for each type of input.
2. **Simplicity:** Developers will be able to define keyboard behavior directly in their models, improving code clarity and maintainability.
3. **Consistency:** It will provide a uniform approach to handling `inputmode` throughout the application.

# Conclusion
I hope this suggestion will be considered for future versions of ASP.NET. I'm open to discussing more about this functionality and getting feedback. Thank you for your attention!

**Comment by huoyaoyuan** [comment]
For numeric fields, the common practice is to define the field in numeric type, instead of string.
For some input types, there are already validation attributes for `[Url]` `[EmailAddress]` etc. A "meta" attribute isn't really helpful since its input mode can't be changed at run time.

**Reply by Andres10121997** [reply]
Thanks for your feedback! I understand that for numeric fields, it is more common to use numeric types instead of strings, and also that validation attributes such as `[Url]` and `[EmailAddress]` already exist. However, my proposal is more focused on improving the usability of forms, particularly on mobile devices.

The goal of the `InputModeAttribute` is to provide developers with a simple way to define what type of keyboard should be displayed in a field, improving the user experience by facilitating correct data entry from a mobile device. Although it is true that it would not be a dynamic attribute, the fact of being able to define the keyboard behavior directly in the model grants clear benefits in terms of accessibility and consistency in the behavior of the forms.

I think it is a complementary functionality to the existing validations, since it focuses on improving the interaction experience, and not only on validating the entered data.

I am open to discuss this topic further and see how we could optimize this idea. Thanks again for your feedback!

**Reply by huoyaoyuan** [reply]
The `ValidationAttribute` itself is about validation and not related to HTML input element. For front-end related discussion you should go to https://github.com/dotnet/aspnetcore repo.

Which HTML application model are you referring to? Razor requires specifying raw HTML elements and Blazor uses different element tags for different input mode. Is there any application model that generates HTML input type based on model type?

**Reply by Andres10121997** [reply]
Thanks again for your reply. I understand your point about the `ValidationAttribute` being more focused on data validation and not on HTML element manipulation. I also appreciate the suggestion to direct these types of discussions to the ASP.NET Core repository.

My proposal is primarily geared towards forms generated via `Razor Pages` or `MVC`, where models define the form fields. The idea is to simplify the assignment of the `inputmode` attribute in forms, making the user experience on mobile devices easier.

I have actually already started a new discussion in the ASP.NET Core repository ([link here](https://github.com/dotnet/aspnetcore/discussions/58036#discussioncomment-10729522)), and there I have been told that Data Annotations live in the dotnet/runtime repository. Therefore, I have reposted here.

Thanks for your feedback and patience as I continue to explore this topic.

**Reply by huoyaoyuan** [reply]
Though the final review and implementation of the type belongs to this repo, the usages in application models like ASP.NET, or languages like C# belongs to their repo.
If there's no direct integration from application models, it doesn't seem useful to be included in the BCL.

**Reply by Andres10121997** [reply]
Thank you for the clarification and for your time. I will continue with this proposal in the discussion I have open in the corresponding repository. I greatly appreciate your feedback and your help in this process.
